src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,733 @@
+			      const char *expect)
+{
+	char *line = tt_static_buf();
+	size_t expect_len = strlen(expect);
+	assert(expect_len < TT_STATIC_BUF_LEN);
+	if (fgets(line, TT_STATIC_BUF_LEN, reader->fp) == NULL ||
+	    memcmp(line, expect, expect_len) != 0) {
+		say_error("%s: file header read failed", reader->filename);
+		return NULL;
+	}
+	return line + expect_len;
+}
+
+struct memtx_sort_data_reader *
+memtx_sort_data_reader_new(const char *snap_filename,
+			   const struct vclock *vclock,
+			   const struct tt_uuid *instance_uuid)
+{
+	const char *filename = memtx_sort_data_filename(snap_filename);
+	struct memtx_sort_data_reader *reader = xmalloc(sizeof(*reader));
+	snprintf(reader->filename, sizeof(reader->filename), "%s", filename);
+	reader->entries = mh_memtx_sort_data_entries_new();
+	reader->curr_entry = NULL;
+	reader->old2new = NULL;
+
+	/* Open the .sortdata file. */
+	reader->fp = fopen(reader->filename, "rb");
+	if (reader->fp == NULL) {
+		say_error("%s: file open failed", reader->filename);
+		goto fail_free;
+	}
+
+	/* Set the read buffer, the PK sort data read is very slow otherwise. */
+	size_t buffer_capacity = 8 * 1024 * 1024; /* 8MB */
+	reader->buffer = xmalloc(buffer_capacity);
+	if (setvbuf(reader->fp, reader->buffer, _IOFBF, buffer_capacity) != 0) {
+		say_error("%s: file buffer set failed", reader->filename);
+		goto fail_close;
+	}
+
+	/* Verify the file magic. */
+	if (memtx_sort_data_header_expect(reader, "SORTDATA\n") == NULL)
+		goto fail_close;
+
+	/* Verify the file version. */
+	if (memtx_sort_data_header_expect(reader, "1\n") == NULL)
+		goto fail_close;
+
+	/* Skip the Tarantool version. */
+	if (memtx_sort_data_header_expect(reader, "Version: ") == NULL)
+		goto fail_close;
+
+	/* Verify the instance UUID. */
+	char *uuid_str = memtx_sort_data_header_expect(reader, "Instance: ");
+	if (uuid_str == NULL)
+		goto fail_close;
+	if (strlen(uuid_str) - 1 /* Newline. */ != UUID_STR_LEN) {
+		say_error("%s: invalid UUID length: %s",
+			  reader->filename, uuid_str);
+		goto fail_close;
+	}
+	uuid_str[UUID_STR_LEN] = '\0';
+	struct tt_uuid sortdata_uuid;
+	if (tt_uuid_from_string(uuid_str, &sortdata_uuid) != 0) {
+		say_error("%s: invalid UUID: %s", reader->filename, uuid_str);
+		goto fail_close;
+	}
+	if (!tt_uuid_is_equal(&sortdata_uuid, instance_uuid)) {
+		say_error("%s: unmatched UUID: %s", reader->filename, uuid_str);
+		goto fail_close;
+	}
+
+	/* Verify the vclock signature. */
+	char *vclock_str = memtx_sort_data_header_expect(reader, "VClock: ");
+	if (vclock_str == NULL)
+		goto fail_close;
+	vclock_str[strlen(vclock_str) - 1] = '\0'; /* Drop the newline. */
+	struct vclock sortdata_vclock = {};
+	if (vclock_from_string(&sortdata_vclock, vclock_str) != 0) {
+		say_error("%s: invalid VClock: %s",
+			  reader->filename, vclock_str);
+		goto fail_close;
+	}
+	long long sortdata_signature = vclock_sum(&sortdata_vclock);
+	long long snapshot_signature = vclock_sum(vclock);
```

Comment:
Let's add a test case with a local space. Just to make sure that local vclock dimension is handled here.